app/Actions/Settings/Login.php

```diff
@@ -53,75 +57,106 @@ public function do(string $username, string $password, ?string $oldUsername, ?st
 			throw new InvalidPropertyException('Could not hash username or password', $e);
 		}
 
-		if (Legacy::SetPassword($hashedUsername, $hashedPassword)) {
+		if ($this->updateAsAdmin($hashedUsername, $hashedPassword, $oldPassword)) {
 			return;
 		}
+		$this->updateAsUser($username, $hashedPassword, $oldPassword, $ip);
+	}
 
-		// > 4.0.8
-		/** @var User $adminUser */
-		$adminUser = User::query()->find(0);
-		if ($adminUser->password === '' && $adminUser->username === '') {
-			$adminUser->username = $hashedUsername;
-			$adminUser->password = $hashedPassword;
-			$adminUser->save();
-			AccessControl::login($adminUser);
+	/**
+	 * Given a hashed username and password, update the Admin credentials.
+	 *
+	 * @param string  $hashedUsername
+	 * @param string  $hashedPassword
+	 * @param ?string $oldPassword
+	 *
+	 * @return bool true if Admin credentials are updated, false otherwise
+	 *
+	 * @throws RuntimeException
+	 * @throws ModelDBException
+	 * @throws InvalidArgumentException
+	 * @throws BadRequestException
+	 * @throws AuthenticationException
+	 * @throws UnauthenticatedException
+	 */
+	private function updateAsAdmin(string $hashedUsername, string $hashedPassword, ?string $oldPassword): bool
+	{
+		if (Authorization::isAdminNotConfigured()) {
+			/** @var User $adminUser */
+			$adminUser = User::query()->findOrFail(0);
+			if ($adminUser->password === '' && $adminUser->username === '') {
```

Comment:
Isn't this condition always true, if the admin is not configured? If the condition is always true, please remove it, otherwise please add a comment to the source code which explains the condition under which `isAdminNotConfigured` equals true, but this condition does not.